opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -118,255 +145,122 @@ message MetricDescriptor {
-    // histogram can go both up and down over time. Recorded values are always >= 0.
-    GAUGE_HISTOGRAM = 3;
-
-    // Integer counter measurement. The value cannot decrease; if value is reset then
-    // start_time_unix_nano should also be reset.
-    // Corresponding values are stored in Int64DataPoint.
-    COUNTER_INT64 = 4;
-
-    // Floating point counter measurement. The value cannot decrease, if
-    // resets then the start_time_unix_nano should also be reset.
-    // Recorded values are always >= 0.
-    // Corresponding values are stored in DoubleDataPoint.
-    COUNTER_DOUBLE = 5;
-
-    // Histogram cumulative measurement.
-    // Corresponding values are stored in HistogramDataPoint. The count and sum of the
-    // histogram cannot decrease; if values are reset then start_time_unix_nano
-    // should also be reset to the new start timestamp.
-    CUMULATIVE_HISTOGRAM = 6;
-
-    // Summary value. Some frameworks implemented Histograms as a summary of observations
-    // (usually things like request durations and response sizes). While it
-    // also provides a total count of observations and a sum of all observed
-    // values, it calculates configurable percentiles over a sliding time
-    // window.
-    // Corresponding values are stored in SummaryDataPoint.
-    SUMMARY = 7;
+    // INSTANTANEOUS is a metric whose values are measured at a particular
+    // instant.
+    //
+    // If an INSTANTANEOUS metric is reported over any time interval the
+    // values are representative of distinct measurements within the
+    // interval.
+    INSTANTANEOUS = 1;
+
+    // DELTA is a metric whose values are measured and successively added
+    // over a reporting interval. The reported value represents the change
+    // from the previous interval.
+    //
+    // Each reported value for a DELTA metric over a time interval is
+    // self-contained, the values do not depend on values in other time
+    // intervals. This is unlike a CUMMULATIVE metric where values are sums
+    // of all measured events in the time interval and all previously
+    // measured values of the metric.
+    //
+    // DELTA metrics are not guaranteed to be monotonic.
+    DELTA = 2;
+
+    // CUMULATIVE is a metric whose values are measured and successively
+    // added to the lifetime total of the metric.
+    //
+    // Each reported value for a CUMULATIVE metric is the sum of all
+    // measurements up to and including that one.
+    //
+    // CUMULATIVE metrics are not guaranteed to be monotonic.
+    CUMULATIVE = 3;
   }
-  Type type = 4;
+  Interval interval = 4;
 
-  // The set of labels associated with the metric descriptor. Labels in this list apply to
-  // all data points.
-  repeated opentelemetry.proto.common.v1.StringKeyValue labels = 5;
-}
+  // monotonic is true if the metric values are entirely non-increasing, or
+  // entirely non-decreasing.
+  bool monotonic = 5;
```

Comment:
@tigrannajaryan I think protobuf go should do this optimization internally. If this is the case probably worth filling a bug to the protobuf library they should pack the members in the best way to minimize memory usage.

I would not try right now to optimize on that.

Reply:
@bogdandrutu in memory representation of a ProtoBuf message is a Go `struct`. If I remember correctly the order of the fields in the struct matches their order in the ProtoBuf declaration. I do not think ProtoBuf generator reorders fields in order to reduce wasted memory due to alignment requirements.
I do not think it should be considered a ProtoBuf generator bug. I would expect that the struct fields preserve the order otherwise readability of the code will be impacted.
This is lower priority issue but at least I would check the memory usage to see how much we are wasting. Perhaps it is negligible.